extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/IncrementalPublishingKafkaIndexTaskRunner.java

```diff
@@ -1504,7 +1505,7 @@ public DateTime getStartTime(@Context final HttpServletRequest req)
     private final int sequenceId;
     private final String sequenceName;
     private final Map<Integer, Long> startOffsets;
-    private final Map<Integer, Long> endOffsets;
+    private final ConcurrentHashMap<Integer, Long> endOffsets;
```

Comment:
Please include a comment that explains two things:

1. why this is a `ConcurrentHashMap`
2. who is allowed to modify it (do you need to be a specific thread or acquire a specific lock)? I bet that if multiple threads modify this concurrently then unexpected things will happen.

Reply:
Well, the problem is that `setEndOffsets()` can be called by both the main thread and the http thread. Actually `setEndOffsets()` also sets `checkpointed` variable too, so I changed to acquire a lock before modifying `endOffsets` and `checkpointed` variables. 

> I bet that if multiple threads modify this concurrently then unexpected things will happen.

I have no idea about this.